### ISSUE TITLE
Modifying Privacy Budget

### ISSUE THREAD

**Issue by mohammadnaseri** [facebookresearch/pytorch-dp#2]
Is it possible to modify epsilon to apply different levels of noise?

**Comment by Darktex** [comment]
Hi @mohammadnaseri !

If I understand your question correctly, you'd like to control epsilon directly as a parameter rather than just have it in output. We are working on a more comprehensive tutorial which will have more context, but here is a TLDR answer: epsilon cannot be set like a normal hyper parameter (eg size of a layer or learning rate), because we need to compute it based on some key parameters (namely how many steps of SGD you do per epoch, how many epochs, what's your clipping threshold and your noise multiplier). It would be very nice to set a privacy budget ahead of time and work on getting the best accuracy for that privacy budget, but unfortunately this is not something we know how to do right now.

That being said,  one thing that you can do is run TF-Privacy's handy [command-line helper](https://github.com/tensorflow/privacy/tree/master/tutorials#using-command-line-interface-for-privacy-budgeting) to quickly try various hyperparameters and check what the resulting epsilon would be. This is quite cheap to run, so you won't need to actually train a model every time.

Once you know the right values for the parameters, then you can kickoff the training having the certainty that you are going to hit the desired epsilon value.

To aid you in your quest, here's some things we'd recommend to get started:

1. As a first-order approximation, try to fix C to 1.0.
2. Try different values for sigma, maybe between 0.5 to 1.5
3. Choose the largest batch size that can fit in your GPU memory
4. Set `alphas=[1 + x / 10.0 for x in range(1, 100)] + list(range(12, 64))`. You can try more if you want, but these should be fine (you are allowed to try however many you want and keep the best resulting epsilon).
5. Remember to set your target delta to a value much smaller than 1/M, where M = number of samples in your data. If you set delta = 1/M, you expect your mechanism to fail for one data point.
6. Try different values for N_epochs starting with 1 and going up until you exceed your privacy budget. Keep the last non-exceeding N.

Once you have 3-5 candidate sets of parameters that will take you to `eps_desired`, simply run a training experiment with each and keep the one with the best accuracy/rocauc/f1/whatever you care about :)

Let us know if this answers your question!

**Comment by mohammadnaseri** [comment]
Hey, thanks for your help, it worked!, I would like to ask if I can apply differential privacy on a trained model as well? I mean if I have a model and before deploying it to production, I add some noise to the model parameters. Does that make sense?

**Comment by ilyamironov** [comment]
I am glad @Darktex's answer helped. As to your second question, namely whether it is possible to add noise to the model after it's been trained, the answer is no, not really. For virtually all learning tasks, we don't know how to estimate the model's sensitivity in respect to the inputs, and thus we don't know provable ways of adding noise required to guarantee differential privacy.

**Comment by AlphaPav** [comment]
Hi @Darktex 

Could we fix an epsilon before training and use the moment accountant to calculate the minimum noise level needed during training? Similar with this function in TensorFlow privacy https://github.com/tensorflow/privacy/blob/ca347b8995cbee924dd121afa941583097ffd3fd/tensorflow_privacy/privacy/analysis/compute_noise_from_budget_lib.py#L48 

Thanks!
